lightning/src/ln/functional_tests.rs

```diff
@@ -2988,6 +2988,77 @@ fn test_commitment_revoked_fail_backward_exhaustive_b() {
 	do_test_commitment_revoked_fail_backward_exhaustive(true, false, true);
 }
 
+#[test]
+fn fail_pending_htlc_backward_upon_channel_failure() {
+	let chanmon_cfgs = create_chanmon_cfgs(3);
+	let node_cfgs = create_node_cfgs(3, &chanmon_cfgs);
+	let node_chanmgrs = create_node_chanmgrs(3, &node_cfgs, &[None, None, None]);
+	let mut nodes = create_network(3, &node_cfgs, &node_chanmgrs);
+	let _chan_1 = create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 1_000_000, 500_000_000, InitFeatures::supported(), InitFeatures::supported());
+	let chan_2 = create_announced_chan_between_nodes_with_value(&nodes, 1, 2, 1_000_000, 500_000_000, InitFeatures::supported(), InitFeatures::supported());
+
+	// Alice -> Bob -> Chuck: Route a payment but without Chuck sending revoke_and_ack.
+	{
+		let (_, payment_hash) = route_payment(&nodes[0], &[&nodes[1]], 50_000);
+		let remaining_route = nodes[1].router.get_route(&nodes[2].node.get_our_node_id(), None, &Vec::new(), 50_000, TEST_FINAL_CLTV).unwrap();
+		nodes[1].node.send_payment(remaining_route, payment_hash).unwrap();
+		check_added_monitors!(nodes[1], 1);
+
+		let payment_event = {
+			let mut events = nodes[1].node.get_and_clear_pending_msg_events();
+			assert_eq!(events.len(), 1);
+			SendEvent::from_event(events.remove(0))
+		};
+		assert_eq!(payment_event.node_id, nodes[2].node.get_our_node_id());
+		assert_eq!(payment_event.msgs.len(), 1);
+	}
+
+	// Alice -> Bob -> Chuck: Route another payment but now Bob waits for Chuck's earlier revoke_and_ack.
+	let (_, failed_payment_hash) = route_payment(&nodes[0], &[&nodes[1]], 50_000);
```

Comment:
Oh lol, guess you dont need three nodes for this test. Note that this is a completely separate payment from the next one - we disambiguate by HTLCSource, not the payment_hash (as otherwise there are a number of privacy and practical funds issues). The payment_failed at the end is the indicator - its saying that a payment nodes[1] tried to send failed, not that it should fail back the HTLC to nodes[0].

Reply:
Groking: what is "the next one" referring to?

Reply:
> That was a reference to the manual send_payment call two lines down.

Ah, I see. This was my attempt at simulating `nodes[2]` not sending `revoke_and_ack`.

If I only need two nodes, then are you saying I can get rid of the `nodes[0]` to `nodes[1]` part entirely (i.e., the places where I'm using `route_payment`)? In that case, what is being "failed backward"?

Reply:
Yep. The part being "failed back" is purely the event telling us that the payment failed (as the lock in question here is taken before we decide if the HTLCSource is an OutboundRoute which we sent or a PreviousHopData which means someone else sent us an HTLC that we relayed).

Reply:
Ok, simplified this by only using two nodes in the test.